fem/tmop.cpp

```diff
@@ -14,12 +14,470 @@
+   // We assume that both A and W are nonsingular.
+
+   auto alpha = det_2D(A);
+   auto omega = W->Elem(0,0)*W->Elem(1,1) - W->Elem(0,1)*W->Elem(1,0);
+   adjoint_2D(A, AdjA);
+   transpose_2D(AdjA, AdjAt);
+
+   mult_aTa_2D(W, WtW);
+   mult_2D(AdjAt, WtW, WRK);
+
+   add_2D(-1.0/omega, A,WRK,WRK2);
+   auto fnorm =  fnorm2_2D(WRK2);
+
+   return 0.25 / (alpha) * fnorm;
+};
+
+template <typename type>
+auto nu011_w( const DenseMatrix *T, const std::vector<type> &W) -> type
+{
+
+   std::vector<type> A;   // T*W = A
+   std::vector<type> AdjA,AdjAt, WtW, WRK, WRK2;
+   mult_2D(T,W,A);
+
+   auto alpha = det_2D(A);
+   auto omega = W[0]*W[3] - W[1]*W[2];
+   adjoint_2D(A, AdjA);
+   transpose_2D(AdjA, AdjAt);
+
+   mult_aTa_2D(W, WtW);
+   mult_2D(AdjAt, WtW, WRK);
+
+   add_2D(-1.0/omega, A,WRK,WRK2);
+   auto fnorm =  fnorm2_2D(WRK2);
+
+   return 0.25 / ( alpha) * fnorm;
+};
+
+template <typename type>
+auto nu014a( const DenseMatrix *W,  std::vector<type> &T) -> type
+{
+   MFEM_VERIFY(W != NULL,
+               "Requires a target Jacobian, use SetTargetJacobian().");
+
+   std::vector<type> A;   // T*W = A
+   mult_2D(T,W,A);
+
+   auto sqalpha = sqrt(det_2D(A));
+   real_t sqomega = pow(W->Det(), 0.5);
+
+   return 0.5*pow(sqalpha/sqomega - sqomega/sqalpha, 2.0);
+};
+
+
+template <typename type>
+auto nu036( const DenseMatrix *W,  std::vector<type> &T) -> type
+{
+   MFEM_VERIFY(W != NULL,
+               "Requires a target Jacobian, use SetTargetJacobian().");
+
+   std::vector<type> A;   // T*W = A
+   std::vector<type> AminusW;  // A-W
+
+   mult_2D(T,W,A);
+   add_2D(-1.0,A,W,AminusW);
+   auto fnorm =  fnorm2_2D(AminusW);
+
+   return 1.0 / (det_2D(A)) * fnorm;
+};
+
+template <typename type>
+auto nu036_w( const DenseMatrix *T, const std::vector<type> &W) -> type
+{
+   std::vector<type> A;   // T*W = A
+   std::vector<type> AminusW;  // A-W
+
+   mult_2D(T,W,A);
+   add_2D(-1.0,A,W,AminusW);
+   auto fnorm =  fnorm2_2D(AminusW);
+
+   return 1.0 / (det_2D(A)) * fnorm;
+};
+
+template <typename type>
+auto nu050( const DenseMatrix *W, const std::vector<type> &T) -> type
+{
+   MFEM_VERIFY(W != NULL,
+               "Requires a target Jacobian, use SetTargetJacobian().");
+   // We assume that both A and W are nonsingular.
+   std::vector<type> A;
+   mult_2D(T,W,A);
+   auto l1_A = sqrt(A[0]*A[0] + A[1]*A[1]);
+   auto l2_A = sqrt(A[2]*A[2] + A[3]*A[3]);
+   auto prod_A = l1_A*l2_A;
+   auto det_A = A[0]*A[3] - A[1]*A[2];
+   auto sin_A = det_A/prod_A;
+   auto cos_A = (A[0]*A[2] + A[1]*A[3])/prod_A;
+
+   auto l1_W = sqrt(W->Elem(0,0)*W->Elem(0,0) + W->Elem(1,0)*W->Elem(1,0));
+   auto l2_W = sqrt(W->Elem(0,1)*W->Elem(0,1) + W->Elem(1,1)*W->Elem(1,1));
+   auto prod_W = l1_W*l2_W;
+   auto det_W = W->Elem(0,0)*W->Elem(1,1) - W->Elem(1,0)*W->Elem(0,1);
+   auto sin_W = det_W/prod_W;
+   auto cos_W = (W->Elem(0,0)*W->Elem(0,1) + W->Elem(1,0)*W->Elem(1,1))/prod_W;
+
+   return (1.0 - cos_A*cos_W - sin_A*sin_W)/(sin_A*sin_W);
+};
+
```

Comment:
What's the difference between `nu050` and `nu050_w`? There's lots of code duplication.
The `nu107a` is reused in both `EvalP` and `AssembleH`, why not 50 (similarly 14 and 51)?